### PR TITLE
fix(): multiple click events issue when clicking a video ad

### PR DESCRIPTION
**Pull request by gunhansancar**

### CODE REVIEW

File: Example/SuperAwesomeExampleTests/UI/Common/AdControllerTests.swift [hunk 1]
```diff
@@ -103,12 +103,4 @@ class AdControllerTests: XCTestCase {
         // Then
         XCTAssertEqual(receivedEvent, AdEvent.adShown)
     }
-
-    func testAdClickedCallback() {
```
Comment (tashmulla) [inline Example/SuperAwesomeExampleTests/UI/Common/AdControllerTests.swift:107]: Any test we can add as a replacement?

File: Pod/Classes/UI/Common/AdController.swift [hunk 2]
```diff
@@ -172,7 +164,7 @@ class AdController: AdControllerType, Injectable {
     }
 
     /// Method that is called when a user clicks / taps on an ad
-    func onAdClicked(_ url: URL) {
+    private func onAdClicked(_ url: URL) {
         logger.success("Event callback: onAdClicked: for url: \(url.absoluteString)")
```
Comment (thomas-james-1986) [inline Pod/Classes/UI/Common/AdController.swift:168]: Is an event actually being sent here? I can only see code for navigation? 🤔

Comment (thomas-james-1986) [inline Pod/Classes/UI/Common/AdController.swift:168]: Ah, I see that this logs `onAdClicked` with the url and the actual event is sent in `navigateToUrl`

File: Pod/Classes/UI/Common/AdController.swift [hunk 3]
```diff
@@ -210,6 +202,8 @@ class AdController: AdControllerType, Injectable {
     }
 
     func handleAdTapForVast() {
+        logger.info("Event callback: adClicked for placement \(placementId)")
```
Comment (thomas-james-1986) [inline Pod/Classes/UI/Common/AdController.swift:205]: Are there any duplicate clicks for VAST ads?

File: Pod/Classes/UI/Video/VideoViewController.swift [hunk 4]
```diff
@@ -119,7 +119,6 @@ import UIKit
     }
 
     private func clickAction() {
-        controller.adClicked()
```
[the remark this answers is not in the document]
Comment (gunhansancar) [reply Pod/Classes/UI/Video/VideoViewController.swift:122]: So we were firing the same `adClicked` event here and also inside the actual `navigateTo` function to follow the link

File: Example/SuperAwesomeExampleTests/UI/Common/AdControllerTests.swift (hunk 1 above)
Comment (gunhansancar) [reply Example/SuperAwesomeExampleTests/UI/Common/AdControllerTests.swift:107]: Adding a necessary test now

File: Pod/Classes/UI/Common/AdController.swift (hunk 3 above)
Comment (gunhansancar) [reply Pod/Classes/UI/Common/AdController.swift:203]: We get the video click through URL from the vast tag. On android we have special case for `CPI` then we use `ad.creative.clickUrl` But I don't know why we don't implement that logic in iOS 🤷

File: Pod/Classes/UI/Common/AdController.swift (hunk 2 above)
Comment (gunhansancar) [reply Pod/Classes/UI/Common/AdController.swift]: This function is renamed to `showBumperIfNeeded` which makes more sense given the functionality of it.

**Comment by gunhansancar** [comment]
Another thing to consider is at what point we should fire the `adClicked` callback.
1- As soon as the `handleAdTap` is called
2- After checking if the URL is valid and showing the Parental gate and Bumper page and sanity check for fast clicks then we fire the `adClicked` event

Currently, I chose the second one but we can also opt for the number one though it could give false click information

File: Example/SuperAwesomeExampleTests/UI/Common/AdControllerTests.swift [hunk 5]
```diff
@@ -106,7 +106,13 @@ class AdControllerTests: XCTestCase {
...
     func testAdClickedCallback() {
         // Given
-        controller.adClicked()
+        let adResponse = AdResponse(123, MockFactory.makeAd(.tag, nil, nil, nil, false, nil, false))
+        adResponse.vast = MockFactory.makeVastAd(clickThrough: "https://www.superawesome.com/")
+        adRepository.response = .success(adResponse)
+        controller.load(123, MockFactory.makeAdRequest())
+
+        // When
+        controller.handleAdTapForVast()
```
Comment (tashmulla) [inline Example/SuperAwesomeExampleTests/UI/Common/AdControllerTests.swift:115]: Nice 👏

**Review by thomas-james-1986 — APPROVED** [review]
Nice one! Side note: Are you seeing VPAID events at the moment? I'm only seeing `adLoaded` on Android on `develop` and my fix branch?

**Comment by gunhansancar** [comment]
> 

Yes, it looks normal on ios

**Pull request merged**